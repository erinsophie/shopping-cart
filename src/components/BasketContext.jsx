import { createContext, useContext, useState, useEffect, useRef } from 'react';

const BasketContext = createContext();

function useBasket() {
  return useContext(BasketContext);
}

function BasketProvider({ children, initialBasket = [] }) {
  const [showBasket, setShowBasket] = useState(false);
  const [basket, setBasket] = useState(initialBasket);
  const basketRef = useRef(basket);

  // when basket changes, set the current ref to be
  // the most recent state of the basket
  // this avoids discrepencies in the quantity
  // of products when addToBasket is clicked in succession

  console.log(basket)

  useEffect(() => {
    basketRef.current = basket;
  }, [basket]);

  function addToBasket(newProduct) {
    const currentBasket = basketRef.current;

    const existingItemIndex = currentBasket.findIndex(
      (item) => item.product.id === newProduct.id,
    );

    if (existingItemIndex !== -1) {
      const updatedBasket = [...currentBasket];
      updatedBasket[existingItemIndex].quantity += 1;
      setBasket(updatedBasket);
    } else {
      setBasket([...currentBasket, { product: newProduct, quantity: 1 }]);
    }
  }

  function updateQuantity(basketItem, event) {
    const currentBasket = basketRef.current;
    const indexOfItem = basket.indexOf(basketItem);
    const updatedBasket = [...currentBasket];

    if (event.target.textContent === '-') {
      if (updatedBasket[indexOfItem].quantity === 1) {
        deleteProduct(basketItem.product.id);
      } else {
        updatedBasket[indexOfItem].quantity -= 1;
        setBasket(updatedBasket);
      }
    } else if (event.target.textContent === '+') {
      updatedBasket[indexOfItem].quantity += 1;
      setBasket(updatedBasket);
    }
  }

  function getSubtotal(basket) {
    const subTotal = basket.reduce(
      (acc, item) => acc + item.product.price * item.quantity,
      0,
    );

    return subTotal.toFixed(2);
  }

  function getBasketAmount() {
    return basket.reduce((acc, item) => acc + item.quantity, 0);
  }

  function deleteProduct(productId) {
    const newBasket = basket.filter((item) => item.product.id !== productId);
    setBasket(newBasket);
  }

  function clearBasket() {
    setShowBasket(false);
    setBasket([]);
  }

  const providerValues = {
    basket,
    setBasket,
    showBasket,
    setShowBasket,
    addToBasket,
    getSubtotal,
    getBasketAmount,
    deleteProduct,
    updateQuantity,
    clearBasket,
  };

  return (
    <BasketContext.Provider value={providerValues}>
      {children}
    </BasketContext.Provider>
  );
}

export { useBasket, BasketProvider };
