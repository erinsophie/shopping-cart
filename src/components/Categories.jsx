import { useState, useEffect } from 'react';
import { capitaliseLetter } from '../utils/utils';
import { Link } from 'react-router-dom';

function Categories() {
  const [categories, setCategories] = useState([]);
  const [loading, setLoading] = useState(true);

  // fetch categories
  useEffect(() => {
    async function fetchCategories() {
      try {
        let response = await fetch(
          'https://fakestoreapi.com/products/categories',
        );
        if (!response.ok) {
          throw new Error('Network response was not ok ' + response.statusText);
        }
        let data = await response.json();
        setCategories(data);
      } catch (error) {
        console.error('Fetch error', error);
      } finally {
        setLoading(false);
      }
    }

    fetchCategories();
  }, []);

  return (
    <div className="flex justify-center gap-7 p-3 md:p-8">
      {loading ? (
        <p>Loading...</p>
      ) : (
        <div className="flex gap-5 md:gap-16">
          {categories.map((category, index) => (
            <Link to={`/${category}`} key={index} aria-label={category}>
              <p className="text-center transform transition-transform ease-in-out duration-200 hover:scale-110">
                {capitaliseLetter(category)}
              </p>
            </Link>
          ))}
        </div>
      )}
    </div>
  );
}

export default Categories;
