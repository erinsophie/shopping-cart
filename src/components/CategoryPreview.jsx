function CategoryPreview({ src, name }) {
  return (
    <div className="flex flex-col gap-2 items-center transform transition-transform ease-in-out duration-200 hover:scale-110 cursor-pointer">
      <div className="w-16 h-16 rounded-full overflow-hidden hover:shadow-xl md:w-24 md:h-24 lg:w-36 lg:h-36">
        <img src={src} alt={name} className="object-cover w-full h-full"></img>
      </div>
      <p className='text-xs text-center lg:text-base'>{name}</p>
    </div>
  );
}

export default CategoryPreview;
