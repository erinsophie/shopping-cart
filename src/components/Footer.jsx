function Footer() {
  return (
    <footer className="border-t border-black font-light text-xs md:text-sm lg:text-base">
      <div className="flex gap-6 p-6 bg-footerBlue md:justify-evenly md:p-10">
        <div>
          <p className="font-medium mb-5">Shop</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Gift cards</li>
            <li className="hover:underline">Sitemap</li>
            <li className="hover:underline">Blog</li>
          </ul>
        </div>
        <div>
          <p className="font-medium mb-5">Sell</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Teams</li>
            <li className="hover:underline">Forums</li>
          </ul>
        </div>
        <div>
          <p className="font-medium mb-5">About</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Policies</li>
            <li className="hover:underline">Investors</li>
            <li className="hover:underline">Careers</li>
            <li className="hover:underline">Press</li>
          </ul>
        </div>
        <div>
          <p className="font-medium mb-5">Help</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Help centre</li>
            <li className="hover:underline">Privacy Settings</li>
          </ul>

          <div className="flex gap-5 text-sm mt-10 md:text-2xl lg:text-3xl">
            <i className="fa-brands fa-instagram"></i>
            <i className="fa-brands fa-facebook"></i>
            <i className="fa-brands fa-pinterest"></i>
            <i className="fa-brands fa-twitter"></i>
            <i className="fa-brands fa-youtube"></i>
          </div>
        </div>
      </div>

      <div className="p-5 flex justify-end bg-footerBlue">
        <p>
          Built by
          <a
            href="https://github.com/erinsophie"
            aria-label="Link to GitHub page"
            target="_blank"
            rel="noopener noreferrer"
            className="hover:underline"
          >
            {' '}
            erinsophie
          </a>{' '}
          |{' '}
          <a
            href="https://github.com/erinsophie/shopping-cart"
            aria-label="Link to GitHub repository"
            target="_blank"
            rel="noopener noreferrer"
            className="hover:underline"
          >
            {' '}
            Source{' '}
          </a>
        </p>
      </div>
    </footer>
  );
}

export default Footer;
