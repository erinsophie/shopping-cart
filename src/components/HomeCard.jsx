import { Link } from 'react-router-dom';
import mug from '../assets/products/mug.jpg';

function HomeCard() {
  return (
    <div className="flex p-8 gap-6 items-center justify-between border border-gray-400 rounded-xl md:gap-12 md:flex-row md:p-10 lg:gap-12 lg:flex-row lg:p-14">
      <div className="flex flex-col gap-2 lg:gap-10">
        <div className="flex flex-col gap-2.5">
          <p className="font-light text-sm text-gray-500">
            Editor&apos;s Picks
          </p>
          <p className="text-base md:text-xl lg:text-2xl">Home Decor</p>
          <Link to="/2">
            <p className="cursor-pointer text-sm md:text-base lg:text-base">
              Shop these unqiue finds{' '}
              <i className="fa-solid fa-arrow-right"></i>
            </p>
          </Link>
        </div>

        <div className="flex flex-col hidden gap-4 text-xs lg:text-base md:block lg:block">
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat.
          </p>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat. Duis aute irure dolor in
            reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
            pariatur.
          </p>
        </div>
      </div>
      <img
        src={mug}
        alt="ceramic mugs"
        className="w-44 rounded-xl cursor-pointer hover:shadow-xl md:w-44 lg:w-64"
      ></img>
    </div>
  );
}

export default HomeCard;
