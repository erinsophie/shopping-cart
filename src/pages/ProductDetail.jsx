import { Link, useParams } from 'react-router-dom';
import categories from '../data/categories';
import allProducts from '../data/products';

function renderStars(count) {
  return Array.from({ length: count }).map((_, index) => (
    <i key={index} className="fa-solid fa-star text-xl"></i>
  ));
}

function ProductDetail() {
  const { categoryId, productId } = useParams();
  const category = categories.find(
    (category) => category.id === parseInt(categoryId),
  );
  const product = allProducts.find(
    (product) => product.id === parseInt(productId),
  );

  // error page if product is not found
  if (!product)
    return (
      <div className="flex flex-col h-screen p-6 gap-3">
        <p className="font-cormorant text-2xl">
          Oops! It seems that this product does not exist
        </p>
        <Link to={`/${categoryId}`} className="text-blue-900 font-bold">
          <i className="fa-solid fa-arrow-left"></i> Go back to {category.name}
        </Link>
      </div>
    );

  return (
    <div className="mt-16 mb-16 ml-32 mr-32">
      <Link to={`/${categoryId}`} className="text-blue-900 font-bold">
        <i className="fa-solid fa-arrow-left"></i> Back to {category.name}
      </Link>

      <div className="flex gap-12 p-24">
        <div className="min-w-[450px] max-w-[450px] min-h-[600px] max-h-[600px]">
          <img src={product.image} className="h-full w-full object-cover"></img>
        </div>

        <div className="flex flex-col gap-6">
          <div className="flex flex-col gap-2">
            <p className="font-cormorant text-3xl">{product.name}</p>
            <p className="text-xl">Â£{product.price}</p>
            <div>{renderStars(product.stars)}</div>
          </div>

          <div className="flex flex-col gap-2">
            <p>Highlights</p>
            <ul className="list-disc ml-4">
              {product.highlights.map((item, index) => (
                <li key={index} className="font-light">
                  {item}
                </li>
              ))}
            </ul>
          </div>

          <div className="flex flex-col gap-2">
            <p>Description</p>
            <p className="font-light">{product.description}</p>
          </div>

          <button className="bg-black text-white rounded-full p-3">
            Add to basket
          </button>

          <div className="flex justify-between items-center">
            <div className="flex flex-col gap-2 items-center">
              <i className="fa-solid fa-truck text-2xl"></i>
              <p>Free shipping</p>
            </div>
            <div className="flex flex-col gap-2 items-center">
              <i className="fa-regular fa-circle-check text-2xl"></i>
              <p>30-Day-Gaurantee</p>
            </div>
            <div className="flex flex-col gap-2 items-center">
              <i className="fa-solid fa-rotate-left text-2xl"></i>
              <p>Simple Returns</p>
            </div>
            <div className="flex flex-col gap-2 items-center">
              <i className="fa-solid fa-gift text-2xl"></i>
              <p>3-5 Day Delivery</p>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default ProductDetail;
