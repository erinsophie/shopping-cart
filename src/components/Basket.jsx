import { useBasket } from './BasketContext';
import { Link } from 'react-router-dom';

function Basket() {
  const {
    basket,
    showBasket,
    setShowBasket,
    getSubtotal,
    getBasketAmount,
    deleteProduct,
    clearBasket,
    updateQuantity,
  } = useBasket();

  return showBasket ? (
    <div className="fixed top-0 right-0 z-10 w-full h-full bg-white p-8 overflow-y-auto border border-black flex flex-col gap-10 md:w-1/2 lg:w-1/3">
      <div className="flex justify-between">
        <div className="flex flex-col gap-2">
          <p>Your basket</p>
          <p data-testid="items-amount">{getBasketAmount()} items</p>
        </div>
        <button
          onClick={() => setShowBasket((prev) => !prev)}
          className="fa-solid fa-xmark text-2xl transform transition-transform ease-in-out duration-200 hover:scale-110"
          data-testid="close-basket-btn"
        ></button>
      </div>

      {getBasketAmount() === 0 ? (
        <div className="flex justify-center">
          <p className="font-cormorant text-2xl">Basket is empty</p>
        </div>
      ) : (
        <div className="flex flex-col gap-4">
          {basket.map((basketItem) => (
            // each basket item
            <div
              key={basketItem.product.id}
              className="flex border border-black"
              data-testid="basket-item"
            >
              <div className="w-32 h-36">
                <img
                  src={basketItem.product.image}
                  alt={basketItem.product.title}
                  className="w-full h-full object-cover"
                ></img>
              </div>

              <div className="flex flex-col justify-between w-full p-4">
                <div className="flex justify-between">
                  <p>{basketItem.product.title}</p>
                  <button
                    onClick={() => deleteProduct(basketItem.product.id)}
                    className="fa-regular fa-trash-can"
                    data-testid={`delete-button-${basketItem.product.id}`}
                  ></button>
                </div>
                <div className="flex justify-between">
                  <p>£{basketItem.product.price}</p>
                  <div className="flex gap-3">
                    <button
                      onClick={(e) => updateQuantity(basketItem, e)}
                      className="bg-gray-300 w-7"
                    >
                      -
                    </button>
                    <p>{basketItem.quantity}</p>
                    <button
                      onClick={(e) => updateQuantity(basketItem, e)}
                      className="bg-gray-300 w-7"
                    >
                      +
                    </button>
                  </div>
                </div>
              </div>
            </div>
          ))}
        </div>
      )}

      <div className="flex justify-between items-center">
        <p data-testid="subtotal">Subtotal: £{getSubtotal(basket)}</p>
        {getBasketAmount() > 0 ? (
          <Link to="/checkout">
            <button
              onClick={clearBasket}
              className="border border-black p-2.5 transform transition-transform ease-in-out duration-200 hover:scale-110"
            >
              Checkout
            </button>
          </Link>
        ) : null}
      </div>
    </div>
  ) : null;
}

export default Basket;
