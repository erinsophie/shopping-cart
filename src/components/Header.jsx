import { useBasket } from '../components/BasketContext';
import { Link } from 'react-router-dom';

function Header() {
  const { showBasket, setShowBasket, getBasketAmount } = useBasket();

  return (
    <header className="bg-purple text-white flex items-center gap-3 p-3 border-b border-gray-300 md:p-6 md:gap-6 md:pl-16 md:pr-16 lg:pl-28 lg:pr-28 lg:p-5 lg:gap-10">
      <div className="flex items-center gap-3 flex-1">
        <Link to="/">
          <h1 className="text-xl md:text-2xl lg:text-3xl">FakeStore</h1>
        </Link>

        <form className="w-full flex gap-3">
          <label htmlFor="searchInput" className="sr-only">
            Search Query
          </label>
          <input
            id="searchInput"
            type="search"
            placeholder="Search for anything"
            className="text-black rounded-full bg-gray-100 w-full p-1 font-light text-xs md:text-sm lg:p-2.5 lg:text-base"
          />
          <button type="submit" aria-label="Search">
            <i className="fa-solid fa-magnifying-glass text-sm lg:text-xl"></i>
          </button>
        </form>
      </div>

      <div className="flex gap-2 items-center transform transition-transform ease-in-out duration-200 hover:scale-110">
        <button
          aria-label="Shopping basket"
          aria-expanded={showBasket}
          data-testid="basket-icon"
          onClick={() => setShowBasket((prev) => !prev)}
        >
          <i className="fa-solid fa-basket-shopping text-sm lg:text-xl"></i>
        </button>

        <p data-testid="basket-quantity">{getBasketAmount()}</p>
      </div>
    </header>
  );
}

export default Header;
