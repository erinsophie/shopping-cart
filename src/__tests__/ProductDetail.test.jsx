import { describe, it, expect, vi, beforeEach } from 'vitest';
import {
  render,
  screen,
  act,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { MemoryRouter, useParams } from 'react-router-dom';
import ProductDetail from '../pages/ProductDetail';
import { BasketProvider } from '../components/BasketContext';

// mock fetch function
window.fetch = vi.fn();

// create mock response
function createFetchResponse(data) {
  return {
    ok: true,
    json: () => Promise.resolve(data),
  };
}

// mock react-router-dom in order to mock useParams
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useParams: vi.fn(),
  };
});

describe('Render tests', () => {
  beforeEach(() => {
    // mock params
    useParams.mockReturnValue({ category: 'jewelery', productId: 5 });
    // set the returned value to be the mock products array
    fetch.mockResolvedValue(createFetchResponse(mockProduct));
  });

  const mockProduct = {
    id: 5,
    title: 'title1',
    price: '129.99',
    category: 'jewelery',
    description: 'lorem ipsum',
    image: 'img',
  };

  let container;

  // snapshot
  it('Matches snapshot', () => {
    const { container: renderedContainer } = render(
      <MemoryRouter>
        <BasketProvider>
          <ProductDetail />
        </BasketProvider>
      </MemoryRouter>,
    );

    container = renderedContainer;
    expect(container).toMatchSnapshot();
  });

  it('Loading text is shown while API is fetching data', async () => {
    render(
      <MemoryRouter>
        <BasketProvider>
          <ProductDetail />
        </BasketProvider>
      </MemoryRouter>,
    );

    const loading = screen.getByText('Loading...');
    expect(loading).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByText('Loading...'));
  });

  it('Makes a GET request to fetch products', () => {
    render(
      <MemoryRouter>
        <BasketProvider>
          <ProductDetail />
        </BasketProvider>
      </MemoryRouter>,
    );

    expect(fetch).toHaveBeenCalledWith('https://fakestoreapi.com/products/5');
  });

  it('Renders correct product details', async () => {
    act(() => {
      render(
        <MemoryRouter>
          <BasketProvider>
            <ProductDetail />
          </BasketProvider>
        </MemoryRouter>,
      );
    });

    expect(await screen.findByText(mockProduct.title)).toBeInTheDocument();
    expect(
      await screen.findByText(mockProduct.description),
    ).toBeInTheDocument();
    expect(await screen.findByText('Back to jewelery')).toBeInTheDocument();
  });
});
