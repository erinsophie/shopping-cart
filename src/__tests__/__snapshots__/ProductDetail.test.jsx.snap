// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render tests > Matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col h-screen p-6 gap-3"
  >
    <p
      class="text-2xl"
    >
      Oops! This product does not exist
    </p>
    <a
      class="text-blue-900 font-bold "
      href="/jewelery"
    >
      <i
        class="fa-solid fa-arrow-left"
      />
       Go back to 
      jewelery
       page
    </a>
  </div>
</div>
`;
